cpu/stm32_common/periph/i2c.c

```diff
@@ -0,0 +1,463 @@
+        if ((i2c->ISR & ERROR_FLAG) || !tick) {
+            /* end transmission */
+            _stop(i2c);
+            return -1;
+        }
+    }
+
+    DEBUG("[i2c] read_regs: send start sequence");
+
+    /* send start sequence and slave address */
+    _start(i2c, address, 1, 0);
+
+    tick = TICK_TIMEOUT;
+    /* wait for ack */
+    while (!(i2c->ISR & I2C_ISR_TXIS) && tick--) {
+        if ((i2c->ISR & ERROR_FLAG) || !tick) {
+            /* end transmission */
+            _stop(i2c);
+            return -1;
+        }
+    }
+
+    DEBUG("[i2c] read_regs: Write register to read\n");
+    i2c->TXDR = reg;
+
+    /* send repeated start sequence, read registers and end transmission */
+    DEBUG("[i2c] read_regs: ACK received, send repeated start sequence\n");
+    return i2c_read_bytes(dev, address, data, length, 0);
+}
+
+int i2c_write_bytes(i2c_t dev, uint16_t address, const void *data,
+                    size_t length, uint8_t flags)
+{
+    assert(dev < I2C_NUMOF);
+
+    I2C_TypeDef *i2c = i2c_config[dev].dev;
+
+    if (!(flags & I2C_NOSTART)) {
+        DEBUG("[i2c] write_bytes: start transmission\n");
+        /* start transmission and send slave address */
+        _start(i2c, address, length, 0);
+    }
+
+    DEBUG("[i2c] write_bytes: write the data\n");
+    /* send out data bytes */
+    if (_write(i2c, data, length) < 0) {
+        DEBUG("[i2c] write_bytes: nothing was written\n");
+        length = 0;
+    }
+
+    if (!(flags & I2C_NOSTOP)) {
+        DEBUG("[i2c] write_bytes: end transmission\n");
+        /* end transmission */
+        _stop(i2c);
+    }
+
+    return length;
+}
+
+int i2c_write_regs(i2c_t dev, uint16_t address, uint16_t reg, const void *data,
+                   size_t length, uint8_t flags)
+{
+    assert(dev < I2C_NUMOF);
+
+    (void)flags;
+    uint16_t tick = TICK_TIMEOUT;
+    I2C_TypeDef *i2c = i2c_config[dev].dev;
+
+    /* Check to see if the bus is busy */
+    while ((i2c->ISR & I2C_ISR_BUSY) && tick--) {
+        if ((i2c->ISR & ERROR_FLAG) || !tick) {
+            return -1;
+        }
+    }
+
+    /* start transmission and send slave address */
+    /* increase length because our data is register+data */
+    _start(i2c, address, length + 1, 0);
+
+    /* send register number */
+    DEBUG("[i2c] write_regs: ACK received, write reg into DR\n");
+    i2c->TXDR = reg;
+
+    /* write out data bytes */
+    if (_write(i2c, data, length) < 0) {
+        length = 0;
+    }
+
+    /* end transmission */
+    _stop(i2c);
+
+    return length;
+}
+
+void i2c_poweron(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+
+    periph_clk_en(i2c_config[dev].bus, i2c_config[dev].rcc_mask);
+}
+
+void i2c_poweroff(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+
+    uint16_t tick = TICK_TIMEOUT;
+
+    while ((i2c_config[dev].dev->ISR & I2C_ISR_BUSY) && tick--) {}
+
+    periph_clk_dis(i2c_config[dev].bus, i2c_config[dev].rcc_mask);
+}
+
+static inline void _start(I2C_TypeDef *i2c, uint8_t address,
```

Comment:
Even though 10-bit address is not yet implemented, I think it would be better to use `uint16_t` for `address`. Maybe pass the flags to be able to enable this feature later on.